[Step1] Kane

kane-young opened the pull request:
@daheenallwhite 

흰!!! 너무 늦은 시간에 첫 PR을 보내게 되네요ㅠ 흰의 주말을 지켜주고 싶어 최대한 빨리 했습니다

## 간략 소개🤚🏼
일단 이번 프로젝트는 저의 첫 MVVM 프로젝트입니다!!
간략하게 프로젝트 전체를 도식화한 그림을 보면
![스크린샷 2021-06-03 오후 10 55 21](https://user-images.githubusercontent.com/64566207/120658659-95beb380-c4c0-11eb-854e-3b7163154e11.png)

`Memo` 는 sample 파일의 json을 파싱하기 위한 Model입니다
그리고 현재는 sample 파일로 view를 채웠지만, 실제 데이터를 운용할 때 `MemoListViewModel`에 저장될 memo 목록을 보내줄 Model을 만들었습니다!

`MemoDetailViewModel`, `MemoListViewModel`, `MemoViewModel`은 각각 
`MemoDetailViewController`, `MemoListViewController` `MemoListCell`에 적용될 ViewModel들입니다

## 미리보는 결과물 📱

### iPad Pro (5th generation)
<img src="https://user-images.githubusercontent.com/64566207/120659455-580e5a80-c4c1-11eb-9bff-8b202fc9495d.gif" width = 500>

<img src="https://user-images.githubusercontent.com/64566207/120659494-62c8ef80-c4c1-11eb-8f6c-39e23f00d607.gif" width = 500>

### iPhone12 Pro Max
<img src="https://user-images.githubusercontent.com/64566207/120659505-65c3e000-c4c1-11eb-89b4-ebcf70d21a80.gif" width = 500>

<img src="https://user-images.githubusercontent.com/64566207/120659508-66f50d00-c4c1-11eb-9dfb-5547a98c3f4c.gif" width = 300>

## 고민했던 점 및 해결 못한 점 🤔
현재 horizontalSizeClass가 .compact 일 경우에는 MemoDetailViewController의 TextView의 background 색이 gray 이며,
.regular 일 경우에는 background 색이 white 입니다!

하지만 화면을 회전할 시, 바로 색이 변하지 않고 cell을 select 했을 시에 변하는데... 이를 어떻게 처리할지 고민입니다!

 ### 문제의 상황
<img src="https://user-images.githubusercontent.com/64566207/120662741-572af800-c4c4-11eb-8304-42cb838253ab.png" width = 500>

### 정상적인 상황
<img src="https://user-images.githubusercontent.com/64566207/120662751-598d5200-c4c4-11eb-9058-2140353277a0.png" width = 500>


이상입니다!!!! 감사합니다!! 😄

kane-young committed:
Feat : Storyboard 삭제 및 초기 화면 설정

kane-young committed:
Feat : Sample Json File 추가 & 대응 Model 구조체 구현

kane-young committed:
Feat : MemoListViewModel에 memos를 제공할 타입 생성

kane-young committed:
Feat : MemoListViewModel 생성
- barButtonItem 을 통해 새로 생성될 추가 버튼에 대응할 addMemo 메서드
- tableView의 numberOfCell 에 대응할 getNumberOfMemo 메서드
- 각 cell에 들어갈 memo 반환하는 getMemo 메서드

kane-young committed:
Feat : Memo List를 나타낼 MemoListViewController 생성

kane-young committed:
Feat : Memo List에 사용될 Cell을 정의

kane-young committed:
Feat : MemoListViewController 를 MemoListCell을 통해서 구현

kane-young committed:
Feat : MemoDetailViewModel 상성 (Memo Detail 화면에 대응)

kane-young committed:
Feat : MemoDetailViewController 생성
- View, Constraints 생성

kane-young committed:
Feat : SplitViewController에 MemoListViewController, MemoDetailViewController 삽입

kane-young committed:
Refactor : 수평방향 sizeclass에 따른 textView 배경 변경

daheenallwhite commented on CloudNotes/CloudNotes/Model/Memo.swift:
```diff
@@ -0,0 +1,28 @@
+  private let lastModified: Int
+  var lastModifiedDate: String {
+    let date = Date(timeIntervalSince1970: TimeInterval(lastModified))
+    let dateFormatter: DateFormatter = DateFormatter()
```
formatted string 을 computed property 로 구현한 이유가 있나요?
lastModified property 는 private 으로 한번 초기화 되면 변동될 일이 없을 것 같아보여요
+) DateFormatter 의 성능이 얼마나 expensive 한지 [관련 블로그 글](https://sarunw.com/posts/how-expensive-is-dateformatter/)을 참고해보세요

kane-young replied:
wow... 엄청 성능이 expensive하군요!! 참고하겠습니다!!

daheenallwhite commented on CloudNotes/CloudNotes/ViewModel/MemoListViewModel.swift:
```diff
@@ -0,0 +1,47 @@
+  private var memoServiceAdapter = MemoProvider()
+  private lazy var memos: [Memo]? = {
+    do {
+      try memoServiceAdapter.setMockData()
+      return self.memoServiceAdapter.getMemo()
```
바로 mockData를 가져오는 방식도 있을텐데 set -> get 하는 이유가 있나요?
Provider 의 `setMockData()` method 는 get & set 을 모두 하는 걸로 보여요

daheenallwhite commented on CloudNotes/CloudNotes/ViewModel/MemoDetailViewModel.swift:
```diff
@@ -0,0 +1,23 @@
+
+final class MemoDetailViewModel {
+  private var memo: Memo?
+  private var title: String { return memo?.title ?? "" }
+  private var body: String { return memo?.body ?? "" }
+  
+  var date: String { return memo?.lastModifiedDate ?? "?" }
+  lazy var content: String = {
+    return title + "\n\n" + body
+  }()
```
빈 문자열이나 '\n\n' 같은 string 객체를 매번 생성하는데 따로 property 로 빼는게 좋지 않을까요?

daheenallwhite commented on CloudNotes/CloudNotes/ViewController/SplitViewController.swift:
```diff
@@ -0,0 +1,26 @@
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    self.delegate = self
+    self.viewControllers = [
+      UINavigationController(rootViewController: MemoListViewController()),
+      UINavigationController(rootViewController: MemoDetailViewController())
+    ]
```
이 SplitViewController 는 지정한 두개의 view controller 만 가질 수 없나요? 이 방식이 확장성에 어떤 영향을 줄까요?

daheenallwhite commented on CloudNotes/CloudNotes/ViewController/MemoListViewController.swift:
```diff
@@ -0,0 +1,69 @@
+  }
+  
+  override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+    guard let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier)
+            as? MemoListCell else {
+      return UITableViewCell()
+    }
+    
+    guard let viewModel = tableViewModel.getMemoViewModel(for: indexPath) else {
+      return UITableViewCell()
+    }
+    
+    cell.configure(with: viewModel)
+    return cell
```
view model & cell 의 configure method 를 사용한 과정이 깔끔하고 가독성도 올라갔네요 👍

daheenallwhite commented on CloudNotes/CloudNotes/ViewController/MemoListViewController.swift:
```diff
@@ -0,0 +1,69 @@
+  // MARK: - Table view Delegate
+  
+  override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+    let navigationController = UINavigationController(rootViewController: MemoDetailViewController())
+    guard let memoDetailVC = navigationController.viewControllers.first as? MemoDetailViewController else {
+      return
+    }
```
1. SplitViewController 의 view controllers 에 이미 MemoDetailViewController 객체가 있을텐데 여기서 왜 새로 생성해주나요?
2. 먼저 memoDetailViewController 를 생성한 다음, 그 property 를 reference 하는 방식 대신에 navigationController 의 first 로 다시 가져오는데 이유가 있나요?

daheenallwhite reviewed (commented):
전반적으로 코드로 UI 구성을 스무스하게 잘 하셨네요
MVVM 을 구현하려는 시도와 도식화한 그림이 이해하는데 도움을 많이 주었어요 👏

생각해 볼 부분
- UISplitViewController 는 container view controller 입니다. 일반적인 view controller 와의 차이점은 무엇이 있을까요?

해결못한 부분에 대한 답변
- 제 시뮬에서는 재현되지 않아서 확인할 수가 없었어요.. 좀 더 자세한 재현 상황을 알려주실 수 있나요?

kane-young committed:
Refactor : orientation 변화에 따른 textView background.color 변화 적용

kane-young committed:
Refactor : DateFormatter에 대한 성능 향상을 위한 로직 이동

kane-young committed:
Style : 반복되는 값 타입 객체 상수 선언 & 상속하지 않을 class final 키워드 추가

kane-young committed:
Refactor : SplitViewController 확장성을 고려해 viewControllers 지정 코드 삭제

kane-young committed:
Refactor : tableView의 didSelect 메서드 내부 변경
- 객체 생성하지 않고 내부 프로퍼티 사용

kane-young committed:
Style : 개행 수정

kane-young commented:
흰!! @daheenallwhite  
빠르고 꼼꼼한 피드백 감사합니다ㅎㅎ  🌈☀️
전체적인 구조 변화는 없습니다

##  생각해 볼 부분에 대한 답변

> UISplitViewController 는 container view controller 입니다. 일반적인 view controller 와의 차이점은 무엇이 있을까요?

**`Container View Controller` 의 경우 Scene을 표현하고 이벤트를 직접 처리하는 역할 대신, 뷰 컨트롤러의 연결 관계를 관리하는 View Controller입니다.** 이들은 내부에 콘텐츠를 배치하는 대신 다른 뷰 컨트롤러를 배치하고, **이들을 서로 유기적인 관계로 엮이도록 만들어 줍니다.** 내비게이션 컨트롤러나 탭 바 컨트롤러, 페이지 컨트롤러, Split 뷰 컨트롤러 등이 대표적인 예입니다. 내비게이션 컨트롤의 경우 Stack구조, 탭 바 컨트롤러의 경우 수평 구조로 뷰 컨트롤러들이 관리되고 있는데, 이번에 사용하게 된 Split View Controller의 경우 뷰 컨트롤러들이 단순한 구조는 아니여서 공부하면서 굉장히 흥미로웠습니다.



> 제 시뮬에서는 재현되지 않아서 확인할 수가 없었어요.. 좀 더 자세한 재현 상황을 알려주실 수 있나요?

iPhone 12 max와 같이 `Landscape orientation` 에서

Regular width, compact height의 Size class를 가진 Device의 경우 아래와 같이 textView background색이 회색으로 유지되는 문제가 있었는데, WWDC 2016 Making Apps Adaptive, Part 1를 보면서 해결 방법을 찾았네요ㅎㅎ

<img width="600" src="https://user-images.githubusercontent.com/64566207/120967449-6bac1080-c7a2-11eb-9087-c57be97269d7.png">

<img width="600" src="https://user-images.githubusercontent.com/64566207/120968038-4c61b300-c7a3-11eb-9dac-821b9c10f184.png">



```swift
override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
   if previousTraitCollection?.horizontalSizeClass == .compact {
     textView.backgroundColor = .white
   } else if previousTraitCollection?.horizontalSizeClass == .regular {
     textView.backgroundColor = .systemGray4
   }
}
```



## 각 Comment에 대한 수정

> formatted string 을 computed property 로 구현한 이유가 있나요?
> lastModified property 는 private 으로 한번 초기화 되면 변동될 일이 없을 것 같아보여요
> +) DateFormatter 의 성능이 얼마나 expensive 한지 [관련 블로그 글](https://sarunw.com/posts/how-expensive-is-dateformatter/)을 참고해보세요

**DateFormatter의 성능에 대한 관련 블로그 글을 읽어보고 효율적인 방법으로 수정했습니다!**

`Memo` 타입의 연산 프로퍼티  -> `MemoListViewModel`에서 DateFormatter 객체를 reuse하는 방법으로 한 번에 lastModified를 String 타입으로 변환 했습니다



>바로 mockData를 가져오는 방식도 있을텐데 set -> get 하는 이유가 있나요?
>Provider 의 `setMockData()` method 는 get & set 을 모두 하는 걸로 보여요

이후 스텝에서 진행할 때 set / get이 분리되어 있어야 편할꺼라고 생각했었는데 다시 생각해보니 그렇네요ㅎㅎ 수정했습니다



> 이 SplitViewController 는 지정한 두개의 view controller 만 가질 수 없나요? 이 방식이 확장성에 어떤 영향을 줄까요?

현재 사용하고 있는 view controller들을 고정으로 두고 이후에 `SplitViewController`를 사용할 때는   `setViewController(_ : for: )` 메서드와` viewControllers` 프로퍼티를 활용해서 변경해줄 수 있지 않을까요? 하지만 이렇게 되면 코드를 직접 보지 않으면 해당 `SplitViewController` 의 primary와 secondary를 알 수 없으니 객체 생성 이후에 지정해주도록 변경하겠습니다!



> 1. SplitViewController 의 view controllers 에 이미 MemoDetailViewController 객체가 있을텐데 여기서 왜 새로 생성해주나요?
> 2. 먼저 memoDetailViewController 를 생성한 다음, 그 property 를 reference 하는 방식 대신에 navigationController 의 first 로 다시 가져오는데 이유가 있나요?

`splitViewController`의 `secondary viewController`를 가져오면 되는데 인지하지 못했네요. 아래와 같이 수정했습니다!!

```swift
guard let splitViewController = splitViewController as? SplitViewController else { 		 
  return
}
guard let detailViewController = splitViewController.viewController(for: .secondary)
            as? MemoDetailViewController else {
  return
}
    
guard let memo = tableViewModel.getMemo(for: indexPath) else { return }
detailViewController.configure(with: memo)
showDetailViewController(detailViewController, sender: self)
```



## 또 다른 수정 사항

`TextView`에 `MemoDetailViewModel`  수정 타이밍을 전달해주기 위해서 `MemoDetailViewModelDelegate` 를 채택해주었습니다

```swift
extension MemoDetailViewController: MemoDetailViewModelDelegate {
  func changeMemo(content: String) {
    textView.text = content
  }
}
```

daheenallwhite commented on CloudNotes/CloudNotes/ViewController/MemoListViewController.swift:
```diff
@@ -0,0 +1,72 @@
+  private let tableViewModel: MemoListViewModel = MemoListViewModel()
+  
+  lazy var addButton: UIBarButtonItem = {
+    let button = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addMemo))
+    
+    return button
```
여기 빈줄은 의미가 있나요?

kane-young replied:
의미 없는 빈줄이네요. 삭제하겠습니다!!!  🤔

daheenallwhite commented on CloudNotes/CloudNotes/ViewModel/MemoDetailViewModel.swift:
```diff
@@ -0,0 +1,38 @@
+  private var memo: Memo = Memo(title: ContentConstant.emptyString,
+                                body: ContentConstant.emptyString,
+                                lastModified: .zero) {
+    willSet {
+      self.date = newValue.lastModified
+      self.content = newValue.title + ContentConstant.doubleNewLine + newValue.body
+      delegate?.changeMemo(content: content)
+    }
```
didSet 이 아닌 willSet 에 구현한 이유가 있나요?

kane-young replied:
newValue를 활용하려고 willSet에 구현을 하였습니다!

kane-young replied:
수정을 하다보니 프로퍼티 감시자가 필요 없겠네요ㅎㅎ
```swift
 private var memo: Memo = Memo(title: ContentConstant.emptyString,
                                body: ContentConstant.emptyString,
                                lastModified: .zero)
  var date: Int { return memo.lastModified }
  var content: String { return memo.title + ContentConstant.doubleNewLine + memo.body }
  
  func configure(with memo: Memo) {
    self.memo = memo
    delegate?.changeMemo(content: content)
  }
```
로 수정하였습니다!!

daheenallwhite commented on CloudNotes/CloudNotes/ViewModel/MemoListViewModel.swift:
```diff
@@ -0,0 +1,60 @@
+    memos?.append(memo)
+  }
+  
+  func getNumberOfMemo() -> Int {
+    guard let memos = memos else { return .zero }
+    
+    return memos.count
```
1. 빈줄에 의미가 있나요?
2. computed property 로 구현하는건 어때요?

kane-young replied:
```swift
var count: Int {
    guard let memos = memos else { return .zero }
    return memos.count
  }
```
위와 같이 수정했습니다!!

daheenallwhite reviewed (commented):
`traitCollectionDidChange` method 를 활용했네요 👍 
Q. `traitCollectionDidChange` 는 언제 호출되나요?
수정된 버전에서는 줄바꿈이 전반적으로 많아졌는데 불필요한 부분이 좀 보여요. 의미가 있는 줄바꿈인지 한번 더 생각해주세요

코멘트 확인해주시고 수정 후에 머지 원하면 알려주세요!

kane-young committed:
Style : 필요 없는 공백 제거

kane-young committed:
Style : MemoDetailViewModel 수정
- memo 프로퍼티 감시자 제거
- date, content 프로퍼티 연산프로퍼티로 수정
- delegate.changeMemo 메서드 configure()로 이동

kane-young committed:
Style : MemoListViewModel memo 갯수 메서드 -> 연산프로퍼티
